### PR TITLE
feat: Add relational filter conditions in data browser

### LINKED ISSUE #2496: Add relational filter conditions
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-dashboard/issues?q=is%3Aissue).

### Current Limitation

The filter dialog in the data browser currently only allows to filter for fields in the current class. It's not possible to add a filter condition that relates to another class.

### Feature / Enhancement Description

Add filter conditions that relate to other classes. Since Parse Dashboard knows the entire schema, it also knows which other classes have a pointer to the currently displayed class; it can use that information to allow for filtering by fields of these related classes. 

The filter dialog could be extended with a drop-down for related classes and its fields.

### Example Use Case

A user in the `_User` class may have
- purchases in a `Purchase` class with a `_User` pointer field
- transactions in a `Transaction` class with a `_User` pointer field

To filter all users who
- made purchases of shoes (in `Purchase` class)
- with a transaction using a credit card (in `Transaction` class)

The filter dialog could look like this:

image

### Alternatives / Workarounds
n/a

### PR DESCRIPTION
**Pull request by vardhan0604**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-dashboard/issues/2496).

### Issue Description

Added filter conditions that relate to other classes. Since Parse Dashboard knows the entire schema, it also knows which other classes have a pointer to the currently displayed class; it can use that information to allow for filtering by fields of these related classes.

image

Closes: #2496

### CODE REVIEW

**Comment by vardhan0604** [comment]
@mtrezza, can you please check this

**Comment by vardhan0604** [comment]
Hey @mtrezza, I have fixed the lint issue, can you please review the PR

**Comment by mtrezza** [comment]
The dashboard data browser seems to be continuously loading the table; not sure if that is a bug related to this feature, still investigating. You can see this by launching the preview environment from the uffizzi-cloud comment and just opening a collection like "_User".

Edit: I can confirm that the issue only occurs in this PR; could you look into that?

File: src/dashboard/Data/Browser/Browser.react.js [hunk 1]
```diff
@@ -101,9 +101,13 @@ class Browser extends DashboardView {
 
       rowCheckboxDragging: false,
       draggedRowSelection: false,
+
+      classes: {},
+      AllclassesSchema: {}
```
Comment (mtrezza) [inline src/dashboard/Data/Browser/Browser.react.js]:
```suggestion
      allClassesSchema: {}
```

File: src/dashboard/Data/Browser/Browser.react.js [hunk 2]
```diff
@@ -101,9 +101,13 @@ class Browser extends DashboardView {
...
     };
 
     this.addLocation = this.addLocation.bind(this);
+    this.AllclassesSchema = this.getAllclassesSchema.bind(this);
```
Comment (mtrezza) [inline src/dashboard/Data/Browser/Browser.react.js]:
```suggestion
    this.allClassesSchema = this.getAllClassesSchema.bind(this);
```

File: src/dashboard/Data/Browser/BrowserToolbar.react.js [hunk 3]
```diff
@@ -225,14 +227,39 @@ const BrowserToolbar = ({
     });
   }
 
+  Allclasses.forEach(className => {
```
Comment (mtrezza) [inline src/dashboard/Data/Browser/BrowserToolbar.react.js]:
```suggestion
  allClasses.forEach(className => {
```

**Review by mtrezza — CHANGES_REQUESTED** [review]
I've marked some var names with incorrect caps. There are more, could you please correct them?

**Comment by vardhan0604** [comment]
> I've marked some var names with incorrect caps. There are more, could you please correct them?

> data browser seems to be continuously loading the table

Sure, I will correct them and I will look into the second issue as well

**Comment by mtrezza** [comment]
The preview environment does not update on every commit. If you want to update the environment you can close and re-open this PR, then after a few minutes uffizzi will post a new comment with the update env link.

**Comment by vardhan0604** [comment]
> The preview environment does not update on every commit. If you want to update the environment you can close and re-open this PR, then after a few minutes uffizzi will post a new comment with the update env link.

ohh okay, I will keep that in my mind

**Comment by vardhan0604** [comment]
> data browser seems to be continuously loading the table

@mtrezza I have fixed this issue and also fixed the var names.
can you please review the code?

File: src/components/BrowserFilter/BrowserFilter.react.js [hunk 4]
```diff
@@ -46,13 +46,18 @@ export default class BrowserFilter extends React.Component {
...
+        this.state.filters
       );
-      const field = Object.keys(available)[0];
-      filters = new List([new Map({ field: field, constraint: available[field][0] })]);
+      const filterClass = Object.keys(available)[0];
+      const filterField = Object.keys(available[filterClass])[0];
+      const filterConstraint = available[filterClass][filterField][0];
+      filters = new List([
+        new Map({ class: filterClass, field: filterField, constraint: filterConstraint }),
+      ]);
```
Comment (mtrezza) [inline src/components/BrowserFilter/BrowserFilter.react.js]: This code block seems to occur almost identical in `addRow`. Could you refactor these occurrences to deduplicate? Maybe keep the `availableFilters` method and move the code block there?

File: src/components/Filter/Filter.react.js [hunk 5]
```diff
@@ -12,28 +12,58 @@ import React from 'react';
...
+
+  const newField = isFieldValid ? field : newClassFields[0];
+  const allowedConstraints = Filters.FieldConstraints[schema[newClassName][newField].type];
+  const isConstraintValid = allowedConstraints.includes(constraint);
+
+  const newConstraint = isConstraintValid ? constraint : allowedConstraints[0];
+
+  const defaultCompare = Filters.DefaultComparisons[schema[newClassName][newField].type];
+
+  const newFilter = new Map({
```
Comment (mtrezza) [inline src/components/Filter/Filter.react.js]:
```suggestion
  const newClassFields = Object.keys(schema[newClassName]);
  const isFieldValid = newClassFields.includes(field);
  const newField = isFieldValid ? field : newClassFields[0];
  const allowedConstraints = Filters.FieldConstraints[schema[newClassName][newField].type];
  const isConstraintValid = allowedConstraints.includes(constraint);
  const newConstraint = isConstraintValid ? constraint : allowedConstraints[0];
  const defaultCompare = Filters.DefaultComparisons[schema[newClassName][newField].type];
  const newFilter = new Map({
```

File: src/components/BrowserFilter/BrowserFilter.react.js (hunk 4 above)
Comment (vardhan0604) [reply src/components/BrowserFilter/BrowserFilter.react.js]: > Maybe keep the availableFilters method

It's still there `findRelatedClasses` uses the availableFilters method

maybe i can make a function like this 
```
export function getFilterDetails(available) {
  const filterClass = Object.keys(available)[0];
  const filterField = Object.keys(available[filterClass])[0];
  const filterConstraint = available[filterClass][filterField][0];
  return { filterClass, filterField, filterConstraint };
}
```

and call it here like

```
addRow() {
    const available = Filters.findRelatedClasses(
      this.props.className,
      this.props.allClassesSchema,
      this.state.blacklistedFilters,
      this.state.filters
    );
    const { filterClass, filterField, filterConstraint } = Filters.getFilterDetails(available);
    this.setState(({ filters }) => ({
      filters: filters.push(
        new Map({ class: filterClass, field: filterField, constraint: filterConstraint })
      ),
      editMode: true,
    }));
  }
  ```
  if you think this works , i can push the changes

Comment (mtrezza) [reply src/components/BrowserFilter/BrowserFilter.react.js]: I haven't looked at the logic; I just notice that there is duplicate code which maybe could be optimized. If your suggestion works and it removes the duplication, please go ahead.

Comment (vardhan0604) [reply src/components/BrowserFilter/BrowserFilter.react.js]: actually the duplicate code is nothing just getting the class , field and constraint value from the available. 

I made a function which takes the available and returns a object which contain of class , field and this makes the code more neat in my take and removes the duplication

**Review by mtrezza — COMMENTED** [review]
Code looks good; will do some test scenarios to see if it breaks anything else.

**Comment by vardhan0604** [comment]
> Code looks good; will do some test scenarios to see if it breaks anything else.

sure lemme know if i need to make any changes

**Comment by mtrezza** [comment]
I tested the [example scenario](https://github.com/parse-community/parse-dashboard/issues/2496#issue-1811823229) described in the feature request and it worked well. I can already tell that this feature will be very helpful to many dashboard users, good job.  Here are just some minor details I have found:

- When adding a new filter condition, the default class should be the currently visible class in the data browser. In my test it defaulted to the custom `Transaction` class despite adding the filter being in the `_User` class.
- For better visual, could you add a margin between the "Clear" and "Add" buttons? This also prevents accidentally clicking "Clear" instead of "Add". FYI, I've added a 150px margin-right to the "Clear" button in the image.
   image
- Could you please add an entry to the README.md under "Features" > "Data Browser" > "Filter" in which you describe in a few words that the data browser allows to add relational filter conditions based on other classes that have a pointer to the current class. Basically, one should understand that a class only shows up if it has a pointer to the class that is currently displayed.

Note:
- I tested saving the filter and applying the saved filter, both worked fine.

**Comment by vardhan0604** [comment]
- >When adding a new filter condition, the default class should be the currently visible class in the data browser. In my test it defaulted to the custom Transaction class despite adding the filter being in the _User class.

- >For better visual, could you add a margin between the "Clear" and "Add" buttons? This also prevents accidentally clicking "Clear" instead of "Add". FYI, I've added a 150px margin-right to the "Clear" button in the image.

- >Could you please add an entry to the README.md under "Features" > "Data Browser" > "Filter" in which you describe in a few words that the data browser allows to add relational filter conditions based on other classes that have a pointer to the current class. Basically, one should understand that a class only shows up if it has a pointer to the class that is currently displayed.


Hey @mtrezza , I have done the first two changes asked by you but for the 3rd I'm not sure where and how to do that?
image
is this how we are supposed to do or how?
sorry i havent contributed in docs before

**Comment by mtrezza** [comment]
Just one last thing, given that our filter dialog is now more complex with an additional column, do you think you could add the headings "Class, Field, Condition, Value", like in the [screenshot](https://github.com/parse-community/parse-dashboard/issues/2496#issue-1811823229)? I believe this would make the UI easier to use. If that's too much work, no worries, we can merge without.

And I found another tiny issue: classes with a pointer to the current class only appear in the filter dialog after reloading the dashboard in the browser window. It seems that the schema is not updated locally after adding a pointer field to a class. I'm not sure whether this is a general dashboard issue that the schema is not updated, or whether it is related to this feature.

**Comment by vardhan0604** [comment]
> Just one last thing, given that our filter dialog is now more complex with an additional column, do you think you could add the headings "Class, Field, Condition, Value", like in the [screenshot](https://github.com/parse-community/parse-dashboard/issues/2496#issue-1811823229)? I believe this would make the UI easier to use. If that's too much work, no worries, we can merge without.
> 
> And I found another tiny issue: classes with a pointer to the current class only appear in the filter dialog after reloading the dashboard in the browser window. It seems that the schema is not updated locally after adding a pointer field to a class. I'm not sure whether this is a general dashboard issue that the schema is not updated, or whether it is related to this feature.

Hey @mtrezza , I have fixed the second issue and added the headings in the UI.  you can check it and test it out.
lemme know if something breaks

**Review by mtrezza — COMMENTED** [review]
It seems that the class is still only displaying after a website refresh. Could you check please? The headings look great!

**Comment by vardhan0604** [comment]
> It seems that the class is still only displaying after a website refresh. Could you check please? The headings look great!

Hey @mtrezza , you can check I have fixed the issue

**Review by mtrezza — DISMISSED** [review]
Looks good! Thanks for this PR!

**Review by mtrezza — APPROVED** [review]

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-alpha.9](https://github.com/parse-community/parse-dashboard/releases/tag/6.0.0-alpha.9)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-beta.2](https://github.com/parse-community/parse-dashboard/releases/tag/6.0.0-beta.2)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0](https://github.com/parse-community/parse-dashboard/releases/tag/6.0.0)

**Pull request merged**